### PR TITLE
[WIP] Bug fix: only remove states when last remote unit departs.

### LINKED ISSUE #1: kube-control.connected state is removed on scale back
As per subject *kube-control.connected* state is removed incorrectly during scale back.

To replicate start with fully deployed k8s bundle and:
````
juju add-unit kubernetes-master
# wait for the unit to deploy and then remove it
juju remove-unit kubernetes-master/1
````

Result:
kubernetes-worker units eventually end up in blocked state because of missing *kube-control.connected* state

Expected result:
*kube-control.connected* state is NOT removed since the relation is still in place

### PR DESCRIPTION
**Pull request by tvansteenburgh**
Fixes #1.

Not tested yet.

### CODE REVIEW

File: provides.py [hunk 1]
```diff
@@ -37,12 +37,12 @@ def joined_or_changed(self):
...
     @hook('{provides:kube-control}-relation-{broken,departed}')
     def departed(self):
-        """Remove all states.
+        """Remove all states when the last remote unit leaves the relation.
 
         """
-        conv = self.conversation()
-        conv.remove_state('{relation_name}.connected')
-        conv.remove_state('{relation_name}.gpu.available')
+        if not self.conversations():
```
Comment (johnsca) [inline provides.py:43]: In an `@hook` context, there will always be exactly one conversation, for the `REMOTE_UNIT` of the hook.  I think that this interface layer is actually correct (save for the `-broken` hook issue mentioned in the other comment) and the issue may be in how it's being used in the charm layer.

File: provides.py [hunk 2]
```diff
@@ -37,12 +37,12 @@ def joined_or_changed(self):
 
     @hook('{provides:kube-control}-relation-{broken,departed}')
```
Comment (johnsca) [inline provides.py:38]: The `-broken` hook completely fails with the current conversation model for interface layers because there is no `REMOTE_UNIT` and that's not handled well.  It also doesn't have a real use in the reactive pattern because you should be removing the states on a per-unit basis anyway and `-broken` just becomes the state being removed from the last conversation and thus starting to match `@when_not` again.

TL;DR: interface layers should only ever use `-departed` and using `-broken` can cause some strange issues.

**Comment by johnsca** [comment]
Closing in favor of #28
